tests/factories.py

```diff
@@ -30,32 +30,42 @@
 
 
 def create_user(
-    username="testuser", workspaces=None, output_checker=False, last_refresh=None
+    username="testuser",
+    workspaces: list[str] | None = None,
+    output_checker=False,
+    last_refresh=None,
 ):
     """Factory to create a user.
 
-    For ease of use, workspaces can either be a list of workspaces, which is
-    converted into an appropriate dict. Or it can be an explicit dict, but must
-    include all expected dict parameters.
+    For ease of use, workspaces is just a flat list of workspace name, which is
+    converted into an appropriate dict with all the right keys.
+
+    If you need to create a more complex workspace dict, you can call
+    create_user_from_dict directly.
     """
-    if isinstance(workspaces, dict):
-        # TODO check this dict is valid?
-        workspaces_dict = workspaces
-    else:
-        if workspaces is None:
-            # default to usual workspace
-            workspaces = ["workspace"]
-
-        workspaces_dict = {
-            workspace: {
-                "project_details": {"name": "project", "ongoing": True},
-                "archived": False,
-            }
-            for workspace in workspaces
+    if workspaces is None:
+        # default to usual workspace
+        workspaces = ["workspace"]
+
+    workspaces_dict = {
+        workspace: {
+            "project_details": {"name": "project", "ongoing": True},
+            "archived": False,
         }
+        for workspace in workspaces
+    }
+
+    return create_user_from_dict(
+        username, workspaces_dict, output_checker, last_refresh
+    )
+
 
+def create_user_from_dict(
+    username, workspaces_dict, output_checker=False, last_refresh=None
```

Comment:
Should these parameters be typed?

Reply:
So, we don't type check tests/ so types annotations are not guaranteed to be 100% accurate, so I've tended not to. I added the list[str] one as documenation to clarify the current confusion.

And the type for workspace_dict is messy, something like `dict[str, bool | dict[str, str | bool]]`

Also, I think this dict mess will go away with proper user model, so I'm inclined to leave it

Reply:
I would like to see this done, but as a separate ticket that enables mypy checking for the entire `tests/` directory & all current typing issues fixed...

Reply:
Might be worth trying. I suspect that it might be quite a bit of work, but may be useful

Resolving this comment for now